libs/header/p2p/server.go

```diff
@@ -201,10 +201,33 @@ func (serv *ExchangeServer[H]) handleRequest(from, to uint64) ([]H, error) {
 	}
 
 	log.Debugw("server: handling headers request", "from", from, "to", to)
+	// check that store has the requested height
 	if !serv.store.HasAt(ctx, to-1) {
-		span.SetStatus(codes.Error, header.ErrNotFound.Error())
-		log.Debugw("server: requested headers not stored", "from", from, "to", to)
-		return nil, header.ErrNotFound
+		head, err := serv.store.Head(ctx)
+		if err != nil {
+			span.SetStatus(codes.Error, err.Error())
+			log.Debugw("server: could not get current head", "err", err)
+			return nil, err
+		}
+
+		// might be a case when store hasn't synced yet to the requested range
+		if uint64(head.Height()) < from {
+			span.SetStatus(codes.Error, header.ErrNotFound.Error())
+			log.Debugw("server: requested headers not stored",
+				"from", from,
+				"to", to,
+				"currentHead",
+				head.Height(),
+			)
+			return nil, header.ErrNotFound
+		}
+
+		log.Debugw("server: change requested range",
+			"prevMaxHeight", to,
+			"newMaxHeight", uint64(head.Height())+1,
+		)
+		// change `to` height to return a partial range
+		to = uint64(head.Height()) + 1
 	}
 
 	headersByRange, err := serv.store.GetRangeByHeight(ctx, from, to)
```

Comment:
Why not make GetRangeByHeight return what it has from the given range?

Reply:
This way, we can just do `if store.Has(from) { GetRangeByHeight } else NotFound`

Reply:
It's not the store's responsibility to change the bounds. It should only handle the request.

Reply:
By covering another case, we could make the store API of getRangeByHeaght slightly more universal. In the end, we would else have less code with the same functionality. Don't think we have to do it in this way right now and keep the way you implemented it. Need to park this in an issue though

Reply:
> By covering another case, we could make the store API of getRangeByHeaght slightly more universal.

Still can't agree with that. It has to have a set of methods that cover basic needs. The store should not make any decisions, imo. We can have a proxy layer - to construct requests in a proper way(universal) - yes

Reply:
Giving a sub-range of requested range is not a decision and can be part of the interface contract and docs

Reply:
No, it's actually a decision, that store is making on its own. I did not request a sub-range, but the full range.

Reply:
Again, the fact that GetRange returns a sub-range can be documented on the interface.

Reply:
> brings complexity

Opposite, it reduces complexity for users. Checking `if err != nil && len(h) != 0` is not complex if appropriately documented. It's also not required to do such a check; if the user does not care about partial ranges, it will just ignore returned headers on ErrNotFound. The power users will check for that when needed, like Exchange. Alternatively, there could be ErrPartialRange that contains partial range, and power users may check for it if needed. 

The complexity comes, exactly, in the current implementation, because of the need to call Has and Head, or "городить огород"

Reply:
> "городить огород"
The same can say about adding additional logic inside the store. That's why I suggested proxy

Reply:
Let's finish a dispute on this point. We have two opposite thoughts and now we start going nowhere

Reply:
I will finish on saying that this огород should live somewhere anyway and it's better be an implementation detail

Reply:
Actually, we should keep our arch clean without огород and костыль

Reply:
Огород in this case is a feature 😂

Reply:
I thought you'd meant "mess" lol